### PR TITLE
feat: table plugin

### PR DESCRIPTION
**Pull request by zoli**
Coming from https://github.com/AppFlowy-IO/AppFlowy/pull/2120.

- New `Node` types 'table' and 'table/cell'. The 'table/cell' node for now contains single `TextNode`. All the data is kept in node attributes. Data related to 'table' like number of rows and columns are in 'table' node attributes and data related to each table cell is in 'table/cell' node attribute (cell height, position of cell its row and column).  The `TextNode` in table cell supports selection menu and toolbar but not all of them work inside the table cell (some of them are not applicable in table cell).
- Added `isApplicable` to `ContextMenuItem` so we can have context-base context menu
- Modified `backspaceEventHandler` so it won't move from one table cell to another cell and also handling backspace when the selection includes part of table cells and none table related nodes.
- Modified `_getNodeInOffset` so selection also works when we have multiple nodes in a single line (same y position). I will add more info on this in another comment.
- Made `ColorPicker` more customizable and reusable for other than toolbar service.

### CODE REVIEW

File: lib/src/render/color_menu/color_picker.dart [hunk 1]
```diff
@@ -105,21 +102,27 @@ class _ColorPickerState extends State<ColorPicker> {
...
       crossAxisAlignment: CrossAxisAlignment.start,
       mainAxisAlignment: MainAxisAlignment.start,
-      children: options
-          .map((e) => _buildColorItem(type, e, e.colorHex == selectedColor))
+      children: colorOptionList.colorOptions
+          .map(
+            (e) => _buildColorItem(
+              colorOptionList.onSubmittedAction,
+              e,
+              e.colorHex == colorOptionList.selectedColorHex,
```
Comment (Xazin) [inline lib/src/render/color_menu/color_picker.dart:114]:
```suggestion
            (colorOption) => _buildColorItem(
              colorOptionList.onSubmittedAction,
              colorOption,
              colorOption.colorHex == colorOptionList.selectedColorHex,
```
Makes it easier to read, as you don't have to "guess" what `e` refers to, the default `e` refers to `element`, which can be anything.

File: lib/src/render/table/table_action.dart [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+  }
+
+  for (var i = row + 1; i < rowsLen; i++) {
+    for (var j = 0; j < colsLen; j++) {
+      transaction.updateNode(getCellNode(tableNode, j, i)!, {
+        'colPosition': j,
+        'rowPosition': i + 1,
+      });
+    }
+  }
```
Comment (Xazin) [inline lib/src/render/table/table_action.dart:148]: I see varying versions of the same nested for loop,

Maybe possible to consolidate it into a private method?

File: lib/src/render/table/table_cell_node_widget.dart [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    return Container(
+      constraints: BoxConstraints(
+        minHeight: context.select((Node n) => n.attributes['height']),
+      ),
+      color: context.select((Node n) {
+        if (n.attributes['backgroundColor'] == null) {
+          return null;
+        }
+        var colorInt = int.tryParse(n.attributes['backgroundColor']);
+        return colorInt != null ? Color(colorInt) : null;
```
Comment (Xazin) [inline lib/src/render/table/table_cell_node_widget.dart:48]:
```suggestion
        final colorInt = int.tryParse(n.attributes['backgroundColor']);
        return colorInt != null ? Color(colorInt) : null;
```
Can be final

File: lib/src/render/table/table_cell_node_widget.dart [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+      }),
+      width: double.infinity,
```
Comment (Xazin) [inline lib/src/render/table/table_cell_node_widget.dart:50]: Is this actually needed?

File: lib/src/render/table/table_col_border.dart [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    return MouseRegion(
+      cursor: SystemMouseCursors.resizeLeftRight,
+      onEnter: (_) => setState(() => _borderHovering = true),
+      onExit: (_) => setState(() => _borderHovering = false),
+      child: GestureDetector(
+        onHorizontalDragStart: (_) => setState(() => _borderDragging = true),
+        onHorizontalDragEnd: (_) => setState(() => _borderDragging = false),
+        onHorizontalDragUpdate: (DragUpdateDetails details) {
+          RenderBox box =
+              _borderKey.currentContext?.findRenderObject() as RenderBox;
```
Comment (Xazin) [inline lib/src/render/table/table_col_border.dart:45]:
```suggestion
          final RenderBox box =
              _borderKey.currentContext?.findRenderObject() as RenderBox;
```
Is not reassigned

File: lib/src/render/table/table_col_border.dart [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+          Offset pos = box.localToGlobal(Offset.zero);
+          double colsHeight = widget.tableNode.colsHeight;
```
Comment (Xazin) [inline lib/src/render/table/table_col_border.dart:47]:
```suggestion
          final Offset pos = box.localToGlobal(Offset.zero);
          final double colsHeight = widget.tableNode.colsHeight;
```
These are not reassigned either as far as I can tell

File: lib/src/render/table/table_col_border.dart [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+          final int direction = details.delta.dx > 0 ? 1 : -1;
+          if ((details.globalPosition.dx - pos.dx - (direction * 90)).abs() >
+                  110 ||
+              (details.globalPosition.dy - pos.dy) > colsHeight + 50 ||
+              (details.globalPosition.dy - pos.dy) < -50) {
+            return;
+          }
+
+          final w = widget.tableNode.getColWidth(widget.colIdx);
```
Comment (Xazin) [inline lib/src/render/table/table_col_border.dart:56]:
```suggestion
          final colWidth = widget.tableNode.getColWidth(widget.colIdx);
```
`w` is very ambiguous

File: lib/src/render/table/table_context_menu_item.dart [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+            pickerItemTextColor:
+                style.selectionMenuItemTextColor ?? Colors.black,
+            colorOptionLists: [
+              ColorOptionList(
+                header: 'background color',
+                selectedColorHex: node.attributes['backgroundColor'],
+                colorOptions: generateBackgroundColorOptions(editorState),
+                onSubmittedAction: (color) {
+                  final transaction = editorState.transaction;
+                  var c = color == node.attributes['backgroundColor']
```
Comment (Xazin) [inline lib/src/render/table/table_context_menu_item.dart:177]:
```suggestion
                  final c = color == node.attributes['backgroundColor']
```
Not sure what `c` should stand for here? maybe use eg. `is` or `has` like `final hasBgColor` for readability

File: lib/src/render/table/table_context_menu_item.dart [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+void _dismissColorMenu() {
+  // workaround: SelectionService has been released after hot reload.
+  final isSelectionDisposed =
+      _editorState?.service.selectionServiceKey.currentState == null;
+  if (isSelectionDisposed) {
+    return;
+  }
+  if (_editorState?.service.selectionService.currentSelection.value == null) {
+    return;
+  }
```
Comment (Xazin) [inline lib/src/render/table/table_context_menu_item.dart:208]:
```suggestion
  if (isSelectionDisposed || _editorState?.service.selectionService.currentSelection.value == null) {
    return;
  }
```
Consolidate

**Comment by zoli** [comment]
Implemented markdown decode/encode feature for table.

**Comment by zoli** [comment]
Just wanted to note the main description of this PR is outdated https://github.com/AppFlowy-IO/appflowy-editor/pull/62#issue-1663045067.

**Comment by LucasXu0** [comment]
@zoli Is it ready to review and merge?

**Comment by zoli** [comment]
> @zoli Is it ready to review and merge?

Yes. The design and maybe some UX corners could improve, But I think enhancing them incrementally is better.

**Comment by annieappflowy** [comment]
I can't wait to try this. When will this feature be released?

**Comment by zoli** [comment]
> I can't wait to try this. When will this feature be released?

In my last discussion with @LucasXu0, he said we should move the plugin to Appflowy repo.

**Pull request merged**